### DISCUSSION TITLE
Ability to turn on/off OpenCL

### DISCUSSION THREAD

**Discussion by force-net** [dlemstra/Magick.NET#684] [Help] [answered]
I use Magick.NET for converting images, I've found that processing some images throws an exception: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'

After some investigation, I can found that I can 'fix' this problem by turning off OpenCL (`OpenCL.IsEnabled = false`)

After another investigation, I found that problem images are very large (13600x10200). So, it seems my GPU cannot convert these images. But because I've found that only big images causes problem and OpenCL conversion is faster, I want to try use OpenCL version for small images and CPU version for big images. Is it possilble?

Does switching OpenCL off/on in runtime will have an effect? Can it cause multi-threaded problems? Is there any better variant to switch image processing engines?

**Comment by dlemstra** [comment]
Can you share a small sample projects that reproduces the access violation? Feel free to contact me by email if you cannot share the image in public.

**Comment by force-net** [comment]
I've tried to reproduce issue and reveal, that error is not stable and does not occur for one concrete file.
Error looks like:
```
Unhandled Exception: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at ImageMagick.MagickImage.NativeMethods.X64.MagickImage_Resize(IntPtr Instance, IntPtr geometry, IntPtr& exception)
   at ImageMagick.MagickImage.NativeMagickImage.Resize(String geometry)
```

Code to reproduce:
```
var fileName1 = @"marbles.jpg";
var fileName2 = @"big_marbles.jpg";
for (var i = 0; i < 10; i++)
{
	Console.WriteLine(i);
	using (var image = new MagickImage(fileName1))
	{
		image.Resize(1280, 960);
		image.Write(fileName1 + ".tmp");
	}
	using (var image = new MagickImage(fileName2))
	{
		image.Resize(1280, 960);
		image.Write(fileName2 + ".tmp");
	}
}
```
I've found that:
1. Processing one big file does not cause exception or crash
2. Multiple processing of one big file does not cause exception or crash
3. Processing different files small and big - causes crash from first or second iteration
4. Processing different small files - seems to be ok

`ImagemagickOpenCLDeviceProfile.xml` contains next lines:

```
<devices>
<device name="CPU" score="7.497"/>
<device platform="Intel(R) OpenCL" vendor="Intel(R) Corporation" name="Intel(R) HD Graphics 4600" version="10.18.14.4251" maxClockFrequency="350" maxComputeUnits="20" score="0.2541"/>
<device platform="Intel(R) OpenCL" vendor="Intel(R) Corporation" name="Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz" version="4.2.0.148" maxClockFrequency="4000" maxComputeUnits="8" score="0.6329"/>
</devices>
```

[files.zip](https://github.com/dlemstra/Magick.NET/files/4842490/files.zip)

**Comment by dlemstra** [comment]
I am unable to reproduce your issue with my hardware. I suspect that this could be related to your setup. And to answer a question you asked earlier. You can turn it on or off but that is a global setting so you could run into multi threading issues.

**Reply by force-net** [reply]
Thank you.
It seems, my issue can relate to dynamic GPU memory (driver allocates and releases RAM as memory for GPU and incorrectly uses it).
Will it work if I switch setting due processing of image, will it works? E.g. I process small image with GPU, turn off OpenCL for new big image, process it without GPU and turn on again. Small image will process correctly, or in there can be racing issue?

**Comment by dlemstra** [comment, marked as answer]
You can safely turn it off and on again. But make sure you don't do that multithreaded because that will cause undefined behavior.
